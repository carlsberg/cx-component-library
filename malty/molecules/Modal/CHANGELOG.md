# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **DSM-471:** added z-index prop to modal overlay ([#508](https://github.com/CarlsbergGBS/cx-component-library/issues/508)) ([14ee034](https://github.com/CarlsbergGBS/cx-component-library/commit/14ee0349b499e6a451e9e9be90b5e420e6fac66e))
- **DSM-687:** Banner breakpoint and stories ([#679](https://github.com/CarlsbergGBS/cx-component-library/issues/679)) ([bcc1536](https://github.com/CarlsbergGBS/cx-component-library/commit/bcc1536734ea1e5a2b36344b51b40e563dab3b23))
- Remove `eslint` warnings and errors ([#322](https://github.com/CarlsbergGBS/cx-component-library/issues/322)) ([4c7bc73](https://github.com/CarlsbergGBS/cx-component-library/commit/4c7bc737de8abb060919ac5ba16220a6d5d0ace9))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- Update components to have a function declaration and remove unneeded fragments ([#502](https://github.com/CarlsbergGBS/cx-component-library/issues/502)) ([b470095](https://github.com/CarlsbergGBS/cx-component-library/commit/b4700954f4e1ae9c7e547f4de37ac2b15a9d55ea))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- **modal:** [DSM-654] Apply theme to the modal buttons ([#659](https://github.com/CarlsbergGBS/cx-component-library/issues/659)) ([8c67191](https://github.com/CarlsbergGBS/cx-component-library/commit/8c67191b8270f686039f3b30714b3e25f9981313))
- **modal:** [DSM-913] add tests ([#826](https://github.com/CarlsbergGBS/cx-component-library/issues/826)) ([28a7937](https://github.com/CarlsbergGBS/cx-component-library/commit/28a79379bf0aeaa908a75f641a1d3aa09593fede))
- New Chip component and minor changes ([#239](https://github.com/CarlsbergGBS/cx-component-library/issues/239)) ([ce92ab0](https://github.com/CarlsbergGBS/cx-component-library/commit/ce92ab0543b96296952a29e2d43d0df29d4e188f))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
- Version bumps for various fixes ([#205](https://github.com/CarlsbergGBS/cx-component-library/issues/205)) ([57c2e30](https://github.com/CarlsbergGBS/cx-component-library/commit/57c2e30804ba0aa6ebdd6a3ac59044b69ad9393f))
