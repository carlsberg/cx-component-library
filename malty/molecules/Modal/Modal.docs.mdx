---
labels: ['react', 'typescript', 'modal']
description: 'A Modal component.'
---
