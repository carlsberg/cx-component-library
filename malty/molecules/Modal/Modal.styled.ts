import styled from 'styled-components';

export const StyledContainer = styled.div`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  align-items: center;
  display: flex;
  justify-content: center;
`;

const resolveSize = (size: string) => {
  let resolvedSize;
  switch (size) {
    case 'Large':
      resolvedSize = '860px';
      break;
    case 'XLarge':
      resolvedSize = '1024px';
      break;
    default:
      // Medium Modal
      resolvedSize = '640px';
      break;
  }
  return resolvedSize;
};

export const StyledModalWrapper = styled.div<{
  size: string;
}>`
  width: 100%;
  max-width: ${({ size }) => resolveSize(size)};
  display: block;
  background-color: ${({ theme }) => theme.colors.colours.default.white.value};
  text-align: center;
  padding: ${({ theme }) => theme.sizes.m.value} 0 ${({ theme }) => theme.sizes.l.value};
  position: relative;
  max-height: 80vh;
  @media (max-width: 768px) {
    max-height: calc(100vh - 16px);
    padding: ${({ theme }) => theme.sizes['2xs'].value} 0 ${({ theme }) => theme.sizes.s.value};
    width: calc(100% - 16px);
  }
`;

export const StyledCloseIconContainer = styled.div`
  cursor: pointer;
  position: absolute;
  top: ${({ theme }) => theme.sizes.m.value};
  right: ${({ theme }) => theme.sizes.m.value};
  @media (max-width: 768px) {
    top: ${({ theme }) => theme.sizes['2xs'].value};
    right: ${({ theme }) => theme.sizes.s.value};
  }
`;

export const StyledTitleContainer = styled.div`
  margin: ${({ theme }) => theme.sizes['2xs'].value} ${({ theme }) => theme.sizes['2xl'].value}
    ${({ theme }) => theme.sizes.xs.value};
  display: inherit;
  @media (max-width: 768px) {
    margin: ${({ theme }) => theme.sizes['2xs'].value} ${({ theme }) => theme.sizes.s.value};
  }
`;

export const StyledContentContainer = styled.div`
  margin: ${({ theme }) => theme.sizes.xs.value} ${({ theme }) => theme.sizes['2xl'].value} 0;
  @media (max-width: 768px) {
    margin: ${({ theme }) => theme.sizes['2xs'].value} ${({ theme }) => theme.sizes.s.value} 0;
  }
`;

export const StyledButtonsWrapper = styled.div`
  margin: ${({ theme }) => theme.sizes['2xl'].value} ${({ theme }) => theme.sizes['2xl'].value} 0;
  display: flex;
  justify-content: center;
  @media (max-width: 768px) {
    flex-direction: column;
    margin: ${({ theme }) => theme.sizes.l.value} ${({ theme }) => theme.sizes.s.value} 0;
  }
`;
export const StyledButtonContainer = styled.div`
  max-width: 264px;
  width: 100%;
  @media (max-width: 768px) {
    flex-direction: column;
    max-width: 100%;
  }
  :first-of-type {
    padding-right: ${({ theme }) => theme.sizes['2xs'].value};
    @media (max-width: 768px) {
      padding-right: 0px;
    }
  }
  :last-of-type {
    padding-left: ${({ theme }) => theme.sizes['2xs'].value};
    @media (max-width: 768px) {
      padding-left: 0px;
      margin-top: 16px;
    }
  }
`;
