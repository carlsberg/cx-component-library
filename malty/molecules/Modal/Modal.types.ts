import { ButtonStyle } from '@carlsberggroup/malty.atoms.button';

export enum ModalSize {
  Medium = 'Medium',
  Large = 'Large',
  XLarge = 'XLarge'
}

export interface ModalProps {
  open: boolean;
  onClose: () => void;
  title?: string;
  dismissible?: boolean;
  whiteBackground?: boolean;
  content: React.ReactNode | JSX.Element;
  size?: ModalSize;
  overlayZindex?: number;
  actions?:
    | {
        variant: ButtonStyle;
        label: string;
        onClick: () => void;
        key?: string;
      }[]
    | React.ReactNode
    | JSX.Element;
}
