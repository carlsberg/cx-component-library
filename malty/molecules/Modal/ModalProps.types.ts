export interface ModalProps {
  showModal: boolean;
  setShowModal: (params: boolean) => void;
}
