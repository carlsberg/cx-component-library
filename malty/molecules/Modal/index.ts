export { Modal } from './Modal';
export { ModalSize } from './Modal.types';
export type { ModalProps } from './Modal.types';
