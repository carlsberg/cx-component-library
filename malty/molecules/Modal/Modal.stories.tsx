import { ButtonStyle } from '@carlsberggroup/malty.atoms.button';
import { IconNamesTypes } from '@carlsberggroup/malty.atoms.icon';
import { Meta, Story } from '@storybook/react';
import React, { useState } from 'react';
import { Modal as ModalComponent } from './Modal';
import { ModalProps } from './Modal.types';

export default {
  title: 'Molecules/Modal',
  component: ModalComponent,
  parameters: {
    importObject: 'Modal',
    importPath: '@carlsberggroup/malty.molecules.modal'
  },
  argTypes: {
    open: {
      control: 'boolean',
      description: 'The bolean state that controlls whether the Modal is open or not'
    },
    setOpen: {
      description: `The useState function that controlls the "open" state`
    },
    title: {
      control: 'text',
      description: 'Modal title'
    },
    text: {
      control: 'text',
      description: 'Modal text'
    },
    icon: {
      control: {
        type: 'select',
        options: IconNamesTypes
      },
      description: 'Main Icon displayed'
    },
    image: {
      control: 'text',
      description: 'Image source url'
    },
    buttons: {
      control: '',
      description: 'An array of maximum 2 buttons '
    }
  }
} as Meta;

const Template: Story<ModalProps> = ({ title, text, icon, image, buttons }: ModalProps) => {
  const [open, setOpen] = useState(true);
  const toggleModal = () => setOpen(!open);
  return (
    <>
      <div style={{ height: '600px', width: '500px' }}>
        <button type="button" onClick={toggleModal}>
          Toggle Modal
        </button>
        <ModalComponent
          open={open}
          setOpen={setOpen}
          title={title}
          text={text}
          icon={icon}
          image={image}
          buttons={buttons}
        />
      </div>
    </>
  );
};

export const Modal = Template.bind({});
Modal.args = {
  title: 'Headline',
  text: `Paragraph block to support main headline(optional)
  And…it can have 2 lines, more than that is just boring…`,
  icon: IconNamesTypes.ItemCheck,
  image: 'https://source.unsplash.com/user/c_v_r/220x280',
  buttons: [
    {
      variant: ButtonStyle.Secondary,
      label: 'Cancel',
      onClick: () => alert('primay button pressed')
    },
    {
      variant: ButtonStyle.Primary,
      label: 'Confirm',
      onClick: () => alert('primay button pressed')
    }
  ]
};
