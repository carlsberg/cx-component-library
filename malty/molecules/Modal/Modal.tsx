import { Button, SizeTypes } from '@carlsberggroup/malty.atoms.button';
import { Icon, IconColors, IconNamesTypes, IconSizesTypes } from '@carlsberggroup/malty.atoms.icon';
import { Overlay } from '@carlsberggroup/malty.atoms.overlay';
import {
  Align as TextAlignType,
  Size as TextSizeType,
  Text,
  Weight as TextWeightType
} from '@carlsberggroup/malty.atoms.text';
import React from 'react';
import {
  StyledButtonContainer,
  StyledButtonsWrapper,
  StyledCloseIconContainer,
  StyledContainer,
  StyledIconContainer,
  StyledImgContainer,
  StyledModalWrapper,
  StyledTextContainer,
  StyledTitleContainer
} from './Modal.styled';
import { ModalProps } from './Modal.types';

export const Modal = ({ open, setOpen, info, buttons, image }: ModalProps) => {
  const closeModal = () => {
    setOpen(false);
  };
  return (
    <>
      {open ? (
        <>
          <Overlay />
          <StyledContainer>
            <StyledModalWrapper>
              <StyledCloseIconContainer onClick={closeModal}>
                <Icon name={IconNamesTypes.Close} size={IconSizesTypes.Large} color={IconColors.Primary} />
              </StyledCloseIconContainer>
              {info.icon && (
                <StyledIconContainer>
                  <Icon name={info.icon} size={IconSizesTypes.Large} color={IconColors.Primary} onClick={closeModal} />
                </StyledIconContainer>
              )}
              <StyledTitleContainer>
                <Text align={TextAlignType.Center} weight={TextWeightType.Bold}>
                  {info.title}
                </Text>
              </StyledTitleContainer>
              <StyledTextContainer>
                <Text align={TextAlignType.Center} size={TextSizeType.MediumSmall}>
                  {info.text}
                </Text>
              </StyledTextContainer>

              {image && (
                <StyledImgContainer>
                  <img src={image} />
                </StyledImgContainer>
              )}

              {buttons && (
                <StyledButtonsWrapper>
                  {buttons.map((btnInstance, index: number) => (
                    <StyledButtonContainer key={index}>
                      <Button
                        fullWidth
                        size={SizeTypes.Large}
                        style={btnInstance.variant}
                        onClick={btnInstance.onClick}
                      >
                        {btnInstance.label}
                      </Button>
                    </StyledButtonContainer>
                  ))}
                </StyledButtonsWrapper>
              )}
            </StyledModalWrapper>
          </StyledContainer>
        </>
      ) : null}
    </>
  );
};
