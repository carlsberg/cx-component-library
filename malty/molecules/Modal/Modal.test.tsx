import { ButtonStyle } from '@carlsberggroup/malty.atoms.button';
import { IconNamesTypes } from '@carlsberggroup/malty.atoms.icon';
import { jsonRenderer } from '@carlsberggroup/malty.utils.test';
import React from 'react';
import { Modal } from './Modal';

describe('molecule modal', () => {
  it('matches snapshot', () => {
    const view = jsonRenderer(
      <Modal
        open
        setOpen={() => false}
        title="Headline"
        text={`Paragraph block to support main headline(optional)
    And…it can have 2 lines, more than that is just boring…`}
        icon={IconNamesTypes.ItemCheck}
        image="http://placehold.it/120x120&text=image1"
        buttons={[
          {
            variant: ButtonStyle.Secondary,
            label: 'Cancel',
            onClick: () => alert('primay button pressed')
          },
          {
            variant: ButtonStyle.Primary,
            label: 'Confirm',
            onClick: () => alert('primay button pressed')
          }
        ]}
      />
    );
    expect(view).toMatchSnapshot();
  });
});
