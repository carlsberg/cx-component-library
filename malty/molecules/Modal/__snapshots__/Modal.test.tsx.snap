// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does molecule modal match snapshot? matches snapshot 1`] = `
Array [
  <div
    className="sc-kDDrLX cxbmWb"
  />,
  <div
    className="sc-iqcoie icYzvN"
  >
    <div
      className="sc-crXcEl kMIvft"
      size="Medium"
    >
      <div
        className="sc-evZas kAxUXz"
        onClick={[Function]}
      >
        <svg
          className="sc-gsnTZi ivMBgY"
          color="digital-black"
          data-testid="svg-component"
          size="24px"
          version="1.1"
          viewBox="0 0 40 40"
          xmlns="http://www.w3.org/2000/svg"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        >
          <g
            fillRule="evenodd"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6.249 4.846l.094.083L12 10.585l5.657-5.656a1 1 0 011.497 1.32l-.083.094L13.415 12l5.656 5.657a1 1 0 01-1.32 1.497l-.094-.083L12 13.415l-5.657 5.656a1 1 0 01-1.497-1.32l.083-.094L10.585 12 4.929 6.343a1 1 0 011.32-1.497z"
            />
          </g>
        </svg>
      </div>
      <div
        className="sc-breuTD irOBpF"
      >
        <span
          className="sc-bczRLJ hROJNc"
        >
          <h1
            className="sc-jqUVSM fkSHeC"
            color="digital-black"
          >
            Headline
          </h1>
        </span>
      </div>
      <div
        className="sc-ksZaOG fVkfGD"
      >
        <p>
          Paragraph block to support main headline(optional) And…it can have 2 lines, more than that is just boring…
        </p>
      </div>
      <div
        className="sc-hAZoDl cYsulU"
      >
        <div
          className="sc-fnykZs jJYqjP"
        >
          <span
            className="sc-bczRLJ hROJNc"
          >
            <button
              className="sc-jSMfEi sc-iBkjds dCcUHP hlHhcU"
              fontSize="16px"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <div
                className="text-container "
              >
                Cancel
              </div>
            </button>
          </span>
        </div>
        <div
          className="sc-fnykZs jJYqjP"
        >
          <span
            className="sc-bczRLJ hROJNc"
          >
            <button
              className="sc-jSMfEi sc-gKXOVf dCcUHP fcGpoe"
              fontSize="16px"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <div
                className="text-container "
              >
                Confirm
              </div>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>,
]
`;
